### ISSUE TITLE
Gem::Version comparisons regressed in 2.7.0 and are no longer transitive

### ISSUE THREAD

**Issue by jhawthorn** [rubygems/rubygems#2595]
While working on [pub_grub](https://github.com/jhawthorn/pub_grub) I came across some strange behaviour in `<=>` (as I was rewriting this method to be faster).

``` ruby
> a, b, c = %w[0.0.beta 0.0.beta.1 0.beta.1].map(&Gem::Version.method(:new))
=> [#<Gem::Version "0.0.beta">, #<Gem::Version "0.0.beta.1">, #<Gem::Version "0.beta.1">]
> a < b
=> true
> b == c
=> true
> c < a
=> true

```

`0.0.beta.1` and `0.beta.1` are considered to be equal (`==`). However, when comparing to `0.0.beta`, `0.beta.1` is lesser (`<`), and `0.0.beta.1` is greater (`>`).

The issue is that `canonical_versions` will strip all `0`s preceding the first "prerelease" (string) segment in the version, whereas the less/greater than comparison doesn't. Because `a < b`, `b == c`, and `c < a`, the transitive property doesn't hold and therefore `Gem::Version`s don't form a partially ordered set (I have _a lot_ of Wikipedia pages open right now :sweat_smile:).

It looks like this behaviour was changed in #1659. Prior to that commit, `0.0.beta.1` was considered greater than `0.beta.1`, and not equal. The previous implementation does not have this problem and IMO is easier to reason about (dropping `.0`s in the middle of a version is confusing).

It looks like the change was intended to have equal gem versions hash equivalently (#1617), not to change `==`, so I think this change was unintentional. FWIW, equal gem versions (even not affected by this) still don't hash equivalently (IMO they shouldn't) because hashes use `eql?` not `==`.

```
> h = {}
=> {}
> h[Gem::Version.new("0.0")] = true
=> true
> h[Gem::Version.new("0.0.0")]
=> nil
> # this is because
> Gem::Version.new("0.0") == Gem::Version.new("0.0.0")
=> true
> Gem::Version.new("0.0").eql?(Gem::Version.new("0.0.0"))
=> false
```

From this the options are:
1. **Revert `==` to the previous behaviour**. `>` and `<` preserve their current behaviour (they've never changed). `==` is different in cases like `0.0.0.alpha == 0.0.alpha`. Since nobody noticed when it changed in 2017, it probably won't be noticed when it's changed back :man_shrugging:.

2. Make `<` and `>` match `==` and use the canonical_versions for comparison. This would fix ordering but I believe would cause issues because requirements like `~> 1.x` would no longer match `1.0.beta`. We would also have three historic behaviours: pre-2016, 2017-2018, and post-2019 :grimacing:.

I strongly feel we should go with option 1, and change `canonical_versions` to match the 2016 behaviour. The most common use cases would remain the same and the equality of versions would be easier to reason about.

In addition (but should be considered separately, this is a complicated enough issue as is) we should consider how hashes work on gem versions. IMO the hash keys should be as precise as possible and `1.0.0` should not hash to `1.0`. If  that behaviour is needed (NB. it doesn't exist now) one can manually use `version.canonical_versions` (we could also create a `#canonicalize` or `#normalize` if it was a common need).

**Comment by segiddins** [comment]
I’ll take a look at this as soon as I can, thanks for the **fantastic** issue writeup!

**Comment by deivid-rodriguez** [comment]
Closed by #2651.
